File: idg/python/kernels.py
import numpy as np
from numba import jit, float32


@jit(nopython=True, fastmath=True, cache=True)
def polyval(coefficients: np.ndarray, x: np.ndarray) -> np.ndarray:
    """
    Numba-compatible polynomial evaluation (equivalent to np.polyval).

    :param coeffs: Polynomial coefficients in descending order [a_n, a_{n-1}, ..., a_0]
    :param x: Value(s) at which to evaluate the polynomial
    :return Array of evaluated polynomial values, shape (len(x))
    """
    result = np.zeros_like(x)
    for i in range(len(x)):
        val = coefficients[0]
        for j in range(1, len(coefficients)):
            val = val * x[i] + coefficients[j]
        result[i] = val
    return result


@jit(nopython=True, fastmath=True, cache=False)
def evaluate_spheroidal(nu: np.ndarray) -> np.ndarray:
    """
    Evaluate the prolate spheroidal wave function.

    param: nu: parameters of the spheroidal wave function
    return: array with value of the spheroidal wave function, shape (len(nu))
    """
    P = np.array(
        [
            [8.203343e-2, -3.644705e-1, 6.278660e-1, -5.335581e-1, 2.312756e-1],
            [4.028559e-3, -3.697768e-2, 1.021332e-1, -1.201436e-1, 6.412774e-2],
        ]
    )
    Q = np.array(
        [
            [1.0000000e0, 8.212018e-1, 2.078043e-1],
            [1.0000000e0, 9.599102e-1, 2.918724e-1],
        ]
    )

    # Create result array
    result = np.zeros_like(nu)

    # Process each part
    for part, end in [(0, 0.75), (1, 1.00)]:
        mask = (nu >= (0.0 if part == 0 else 0.75)) & (nu <= end)
        if not np.any(mask):
            continue

        nu_part = nu[mask]
        nusq = nu_part**2
        delnusq = nusq - end**2

        # Calculate polynomial using Horner's method
        top = polyval(P[part][::-1], delnusq)
        bot = polyval(Q[part][::-1], delnusq)

        # Avoid division by zero
        valid = bot != 0
        result_part = np.zeros_like(nu_part)
        result_part[valid] = (1.0 - nusq[valid]) * (top[valid] / bot[valid])
        result[mask] = result_part

    return result


@jit(nopython=True, fastmath=True, cache=True, nogil=True)
def add_pt_src_to_baseline(
    bl: int,  # baseline index
    nr_time: int,  # number of timesteps
    nr_channels: int,  # number of frequency channels
    amplitude: complex,  # amplitude of the point source
    frequencies: np.ndarray,  # frequencies of the channels
    uvw: np.ndarray,  # uvw coordinates
    l: float,  # l coordinate of the point source
    m: float,  # m coordinate of the point source
    visibilities: np.ndarray,  # visibility data
) -> None:
    """
    Update the visibility data for a single baseline with a point source.

    :param bl: baseline index
    :param nr_time: number of timesteps
    :param nr_channels: number of frequency channels
    :param amplitude: amplitude of the point source
    :param frequencies: frequencies of the channels
    :param uvw: uvw coordinates
    :param l: l coordinate of the point source
    :param m: m coordinate of the point source
    :param visibilities: visibility array, shape (nr_baselines, nr_time, nr_channels, nr_correlations)
    """
    speed_of_light = 299792458.0

    for t in range(nr_time):
        for c in range(nr_channels):
            u = (frequencies[c] / speed_of_light) * uvw["u"][bl, t]
            v = (frequencies[c] / speed_of_light) * uvw["v"][bl, t]

            phase = -2 * np.pi * (u * l + v * m)
            value = amplitude * np.exp(1j * phase)

            visibilities[bl, t, c, :] += value


@jit(nopython=True, fastmath=False, cache=True, nogil=True)
def visibilities_to_subgrid(
    s: int,
    metadata: np.ndarray,
    w_step: float,
    grid_size: int,
    image_size: float,
    wavenumbers: np.ndarray,
    visibilities: np.ndarray,
    uvw: np.ndarray,
    taper: np.ndarray,
    nr_correlations_in: int,
    subgrid_size: int,
    subgrid: np.ndarray,
) -> None:
    """
    Grid visibilities onto a subgrid.

    :param s: subgrid index
    :param metadata: metadata for the subgrid
    :param w_step: w step in wavelengths
    :param grid_size: grid size in pixels
    :param image_size: image size in radians
    :param wavenumbers: wavenumbers of the frequencies
    :param visibilities: visibility data
    :param uvw: uvw coordinates
    :param taper: taper function
    :param nr_correlations_in: number of input correlations
    :param subgrid_size: subgrid size in pixels
    :param subgrid: subgrid array
    """
    # Load metadata
    m = metadata[s]
    bl = m["baseline"]
    offset = m["time_index"]
    nr_timesteps = m["nr_timesteps"]
    channel_begin = m["channel_begin"]
    channel_end = m["channel_end"]
    x_coordinate = m["coordinate"]["x"]
    y_coordinate = m["coordinate"]["y"]
    w_offset_in_lambda = w_step * (m["coordinate"]["z"] + 0.5)
    nr_correlations_out = 4 if nr_correlations_in == 4 else 1

    # Compute offsets
    u_offset = (x_coordinate + subgrid_size / 2 - grid_size / 2) * (
        2 * np.pi / image_size
    )
    v_offset = (y_coordinate + subgrid_size / 2 - grid_size / 2) * (
        2 * np.pi / image_size
    )
    w_offset = 2 * np.pi * w_offset_in_lambda

    # Precompute constants
    half_subgrid = subgrid_size / 2
    image_scale = image_size / subgrid_size

    for y in range(subgrid_size):
        for x in range(subgrid_size):

            # Compute l, m, n
            l = float32((x + 0.5 - half_subgrid) * image_scale)
            m = float32((y + 0.5 - half_subgrid) * image_scale)
            tmp = float32(l * l + m * m)  # type: ignore
            n = float32(tmp / (1.0 + np.sqrt(1.0 - tmp)))  # type: ignore

            # Compute pixels
            pixels = np.zeros(nr_correlations_out, dtype=np.complex64)

            for time in range(nr_timesteps):
                idx = offset + time
                u = uvw["u"][bl][idx]
                v = uvw["v"][bl][idx]
                w = uvw["w"][bl][idx]

                phase_index = float32(u * l + v * m + w * n)
                phase_offset = float32(u_offset * l + v_offset * m + w_offset * n)

                for chan in range(channel_begin, channel_end):
                    phase = float32(phase_offset - (phase_index * wavenumbers[chan]))
                    phasor = np.exp(1j * phase)  # type: ignore

                    for pol in range(nr_correlations_in):
                        pixels[pol % nr_correlations_out] += (
                            visibilities[bl, idx, chan, pol] * phasor
                        )

            # Apply taper and store
            sph = taper[y, x]
            x_dst = int((x + half_subgrid) % subgrid_size)
            y_dst = int((y + half_subgrid) % subgrid_size)

            for pol in range(nr_correlations_out):
                subgrid[pol, y_dst, x_dst] = pixels[pol] * sph


@jit(nopython=True, fastmath=True, cache=True)
def compute_phasor(subgrid_size: int) -> np.ndarray:
    """
    Compute the phasor which is used to shift the subgrid to the correct position
    in the grid.

    :param subgrid_size: size of the subgrid
    :return: phasor array, shape (subgrid_size, subgrid_size)
    """
    phasor = np.zeros(shape=(subgrid_size, subgrid_size), dtype=np.complex64)
    for y in range(subgrid_size):
        for x in range(subgrid_size):
            phase = np.float32(np.pi * (x + y - subgrid_size) / subgrid_size)
            phasor[y, x] = np.exp(1j * phase)
    return phasor


@jit(nopython=True, fastmath=True, cache=True)
def add_subgrid_to_grid(
    s: int,
    metadata: np.ndarray,
    subgrids: np.ndarray,
    grid: np.ndarray,
    phasor: np.ndarray,
    nr_correlations: int,
    subgrid_size: int,
    grid_size: int,
) -> None:
    """
    Add a subgrid to the grid.

    :param s: subgrid index
    :param metadata: metadata array
    :param subgrids: subgrid array
    :param grid: grid array
    :param phasor: phasor array
    :param nr_correlations: number of correlations
    :param subgrid_size: size of the subgrid
    :param grid_size: size of the grid
    """
    # Load metadata
    m = metadata[s]

    # Load position in grid
    coordinate = m["coordinate"]
    grid_x = coordinate["x"]
    grid_y = coordinate["y"]

    # Check whether subgrid fits in grid
    if (
        grid_x >= 0
        and grid_x < grid_size - subgrid_size
        and grid_y >= 0
        and grid_y < grid_size - subgrid_size
    ):
        for y in range(subgrid_size):
            for x in range(subgrid_size):
                # Compute shifted position in subgrid
                x_src = int((x + (subgrid_size / 2)) % subgrid_size)
                y_src = int((y + (subgrid_size / 2)) % subgrid_size)

                # Add subgrid value to grid
                for p in range(nr_correlations):
                    grid[p, grid_y + y, grid_x + x] += np.complex64(
                        subgrids[s, p, y_src, x_src] * phasor[y, x]
                    )


@jit(nopython=True, fastmath=True, cache=True)
def compute_metadata(
    grid_size: int,
    subgrid_size: int,
    nr_channels: int,
    bl: int,
    u_pixels: np.ndarray,
    v_pixels: np.ndarray,
    max_group_size: int,
) -> list:
    """
    Compute the metadata for a given baseline.

    :param grid_size: size of the grid
    :param subgrid_size: size of the subgrid
    :param nr_channels: number of frequency channels
    :param bl: baseline index
    :param u_pixels: U coordinates of the visibilities
    :param v_pixels: V coordinates of the visibilities
    :param max_group_size: maximum number of visibilities (timesteps) in a group
    :return: metadata array, shape (nr_subgrids)
    """
    metadata = []

    nr_timesteps = u_pixels.shape[0]
    max_distance = 0.8 * subgrid_size

    # Iterate all timesteps
    timestep = 0
    while timestep < nr_timesteps:
        # Start a new group
        current_u = u_pixels[timestep]
        current_v = v_pixels[timestep]

        # Find consecutive timesteps that are close enough
        group_size = 1
        while (
            timestep + group_size < nr_timesteps
            and group_size < max_group_size
            and np.sqrt(
                (u_pixels[timestep + group_size] - current_u) ** 2
                + (v_pixels[timestep + group_size] - current_v) ** 2
            )
            <= max_distance
        ):
            group_size += 1

        # Calculate group center and subgrid coordinate
        group_u = np.mean(u_pixels[timestep : timestep + group_size])
        group_v = np.mean(v_pixels[timestep : timestep + group_size])

        subgrid_x = int(group_u - (subgrid_size / 2))
        subgrid_y = int(group_v - (subgrid_size / 2))
        subgrid_x = max(0, min(grid_size - subgrid_size, subgrid_x))
        subgrid_y = max(0, min(grid_size - subgrid_size, subgrid_y))

        # Create metadata
        metadata.append(
            (
                bl,
                timestep,
                group_size,
                0,
                nr_channels,
                (subgrid_x, subgrid_y, 0),
            ),
        )
        timestep += group_size

    return metadata

File: idg/cpp/src/kernels.cpp
#include <algorithm>
#include <cmath>
#include <complex>
#include <vector>

#include <xtensor/containers/xarray.hpp>
#include <xtensor/core/xmath.hpp>
#include <xtensor/core/xshape.hpp>
#include <xtensor/views/xindex_view.hpp>
#include <xtensor/views/xview.hpp>

#include "idgtypes.h"

void add_pt_src_to_baseline(size_t bl, size_t nr_timesteps, size_t nr_channels,
                            std::complex<float> amplitude,
                            const xt::xarray<float> &frequencies,
                            const xt::xarray<UVW> &uvw, float l, float m,
                            xt::xarray<std::complex<float>> &visibilities) {
  const float speed_of_light = 299792458.0f;

  for (int t = 0; t < nr_timesteps; ++t) {
    for (int c = 0; c < nr_channels; ++c) {
      float u = (frequencies(c) / speed_of_light) * uvw(bl, t).u;
      float v = (frequencies(c) / speed_of_light) * uvw(bl, t).v;

      float phase = -2.0f * M_PI * (u * l + v * m);
      std::complex<float> value =
          amplitude * std::exp(std::complex<float>(0.0f, phase));

      // Add to all correlations
      for (int pol = 0; pol < visibilities.shape(3); ++pol) {
        visibilities(bl, t, c, pol) += value;
      }
    }
  }
}

// Polynomial evaluation (equivalent to np.polyval)
xt::xarray<float> polyval(const xt::xarray<float> &coefficients,
                          const xt::xarray<float> &x) {
  xt::xarray<float> result = xt::zeros_like(x);

  for (size_t i = 0; i < x.size(); ++i) {
    float val = coefficients(0);
    for (size_t j = 1; j < coefficients.size(); ++j) {
      val = val * x(i) + coefficients(j);
    }
    result(i) = val;
  }
  return result;
}

xt::xarray<float> evaluate_spheroidal(const xt::xarray<float> &nu) {
  xt::xarray<float> P = {
      {8.203343e-2f, -3.644705e-1f, 6.278660e-1f, -5.335581e-1f, 2.312756e-1f},
      {4.028559e-3f, -3.697768e-2f, 1.021332e-1f, -1.201436e-1f, 6.412774e-2f}};

  xt::xarray<float> Q = {{1.0000000e0f, 8.212018e-1f, 2.078043e-1f},
                         {1.0000000e0f, 9.599102e-1f, 2.918724e-1f}};

  xt::xarray<float> result = xt::zeros_like(nu);

  // Process each part
  for (size_t part = 0; part < 2; ++part) {
    float end = (part == 0) ? 0.75f : 1.00f;
    float start = (part == 0) ? 0.0f : 0.75f;

    // Create mask using XTensor's element-wise operations
    xt::xarray<bool> mask = (nu >= start) && (nu <= end);

    // Check if any elements in mask are true
    if (xt::sum(mask)() == 0) {
      continue;
    }

    // Use xt::filter to extract elements
    xt::xarray<float> nu_part = xt::filter(nu, mask);
    xt::xarray<float> nusq = xt::square(nu_part);
    xt::xarray<float> delnusq = nusq - end * end;

    // Reverse coefficients for polyval
    xt::xarray<float> P_part = xt::view(P, part, xt::all());
    xt::xarray<float> Q_part = xt::view(Q, part, xt::all());
    xt::xarray<float> P_reversed = xt::flip(P_part, 0);
    xt::xarray<float> Q_reversed = xt::flip(Q_part, 0);

    // Calculate polynomial using Horner's method
    xt::xarray<float> top = polyval(P_reversed, delnusq);
    xt::xarray<float> bot = polyval(Q_reversed, delnusq);

    // Create valid mask using element-wise comparison
    xt::xarray<bool> valid = xt::not_equal(bot, 0.0f);

    // Calculate result_part
    xt::xarray<float> result_part = xt::zeros_like(nu_part);

    // Use xt::filter to apply the valid mask
    xt::xarray<float> valid_top = xt::filter(top, valid);
    xt::xarray<float> valid_bot = xt::filter(bot, valid);
    xt::xarray<float> valid_nusq = xt::filter(nusq, valid);

    xt::xarray<float> valid_result =
        (1.0f - valid_nusq) * (valid_top / valid_bot);

    // Assign the valid results back to result_part
    xt::xarray<bool> valid_indices = xt::arange(nu_part.size());
    valid_indices = xt::filter(valid_indices, valid);

    for (size_t i = 0; i < valid_indices.size(); ++i) {
      const size_t idx = valid_indices(i);
      result_part(idx) = valid_result(i);
    }

    // Assign result_part back to the original result array
    xt::xarray<bool> mask_indices = xt::arange(nu.size());
    mask_indices = xt::filter(mask_indices, mask);

    for (size_t i = 0; i < mask_indices.size(); ++i) {
      const size_t idx = mask_indices(i);
      result(idx) = result_part(i);
    }
  }

  return result;
}

xt::xarray<std::complex<float>> compute_phasor(size_t subgrid_size) {
  xt::xarray<std::complex<float>> phasor =
      xt::zeros<std::complex<float>>({subgrid_size, subgrid_size});

  for (size_t y = 0; y < subgrid_size; ++y) {
    for (size_t x = 0; x < subgrid_size; ++x) {
      const float phase = M_PI * (x + y - subgrid_size) / subgrid_size;
      phasor(y, x) = std::exp(std::complex<float>(0.0f, phase));
    }
  }

  return phasor;
}

void add_subgrid_to_grid(size_t s, const std::vector<Metadata> &metadata,
                         const xt::xarray<std::complex<float>> &subgrids,
                         xt::xarray<std::complex<float>> &grid,
                         const xt::xarray<std::complex<float>> &phasor,
                         size_t nr_correlations, size_t subgrid_size,
                         size_t grid_size) {
  const Metadata &m = metadata[s];
  const size_t grid_x = m.coordinate.x;
  const size_t grid_y = m.coordinate.y;

  if (grid_x >= 0 && grid_x <= grid_size - subgrid_size && grid_y >= 0 &&
      grid_y <= grid_size - subgrid_size) {

    for (size_t y = 0; y < subgrid_size; ++y) {
      for (size_t x = 0; x < subgrid_size; ++x) {
        const size_t x_src = (x + subgrid_size / 2) % subgrid_size;
        const size_t y_src = (y + subgrid_size / 2) % subgrid_size;

        for (size_t p = 0; p < nr_correlations; ++p) {
          grid(p, grid_y + y, grid_x + x) +=
              subgrids(s, p, y_src, x_src) * phasor(y, x);
        }
      }
    }
  }
}

std::vector<Metadata> compute_metadata(size_t grid_size, size_t subgrid_size,
                                       size_t nr_channels, size_t bl,
                                       const xt::xarray<float> &u_pixels,
                                       const xt::xarray<float> &v_pixels,
                                       size_t max_group_size) {
  std::vector<Metadata> metadata;
  const size_t nr_timesteps = u_pixels.size();
  const float max_distance = 0.8f * subgrid_size;

  size_t timestep = 0;
  while (timestep < nr_timesteps) {
    const float current_u = u_pixels(timestep);
    const float current_v = v_pixels(timestep);

    size_t group_size = 1;
    while (timestep + group_size < nr_timesteps &&
           group_size < max_group_size) {
      const float u_diff = u_pixels(timestep + group_size) - current_u;
      const float v_diff = v_pixels(timestep + group_size) - current_v;
      const float distance = std::sqrt(u_diff * u_diff + v_diff * v_diff);

      if (distance > max_distance) {
        break;
      }
      group_size++;
    }

    // Calculate group center
    float group_u = 0.0f;
    float group_v = 0.0f;
    for (size_t i = 0; i < group_size; ++i) {
      group_u += u_pixels(timestep + i);
      group_v += v_pixels(timestep + i);
    }
    group_u /= group_size;
    group_v /= group_size;

    size_t subgrid_x = group_u - subgrid_size / 2;
    size_t subgrid_y = group_v - subgrid_size / 2;
    subgrid_x = std::max(0ul, std::min(grid_size - subgrid_size, subgrid_x));
    subgrid_y = std::max(0ul, std::min(grid_size - subgrid_size, subgrid_y));

    const Metadata entry{
        static_cast<int>(bl),
        static_cast<int>(timestep),
        static_cast<int>(group_size),
        0,
        static_cast<int>(nr_channels),
        {static_cast<int>(subgrid_x), static_cast<int>(subgrid_y), 0}};

    metadata.push_back(entry);
    timestep += group_size;
  }

  return metadata;
}

void visibilities_to_subgrid(size_t s, const std::vector<Metadata> &metadata,
                             float w_step, size_t grid_size, float image_size,
                             const xt::xarray<float> &wavenumbers,
                             const xt::xarray<VisibilityType> &visibilities,
                             const xt::xarray<float> &uvw_u,
                             const xt::xarray<float> &uvw_v,
                             const xt::xarray<float> &uvw_w,
                             const xt::xarray<float> &taper,
                             size_t nr_correlations_in, size_t subgrid_size,
                             xt::xarray<std::complex<float>> &subgrid) {
  const Metadata &m = metadata[s];
  const int bl = m.baseline;
  const int offset = m.time_index;
  const int nr_timesteps = m.nr_timesteps;
  const int channel_begin = m.channel_begin;
  const int channel_end = m.channel_end;
  const int x_coordinate = m.coordinate.x;
  const int y_coordinate = m.coordinate.y;
  const float w_offset_in_lambda = w_step * (m.coordinate.z + 0.5f);

  const size_t nr_correlations_out = (nr_correlations_in == 4) ? 4 : 1;

  // Compute offsets
  const float u_offset =
      (x_coordinate + subgrid_size / 2.0f - grid_size / 2.0f) *
      (2.0f * M_PI / image_size);
  const float v_offset =
      (y_coordinate + subgrid_size / 2.0f - grid_size / 2.0f) *
      (2.0f * M_PI / image_size);
  const float w_offset = 2.0f * M_PI * w_offset_in_lambda;

  const int half_subgrid = subgrid_size / 2;
  const float image_scale = image_size / subgrid_size;

  for (size_t y = 0; y < subgrid_size; ++y) {
    for (size_t x = 0; x < subgrid_size; ++x) {
      // Compute l, m, n
      const float l = (x + 0.5f - half_subgrid) * image_scale;
      const float m_val = (y + 0.5f - half_subgrid) * image_scale;
      const float tmp = l * l + m_val * m_val;
      const float n = tmp / (1.0f + std::sqrt(1.0f - tmp));

      // Initialize pixels for each correlation
      std::vector<std::complex<float>> pixels(nr_correlations_out, 0.0f);

      for (size_t time = 0; time < nr_timesteps; ++time) {
        const size_t idx = offset + time;
        const float u = uvw_u(bl, idx);
        const float v = uvw_v(bl, idx);
        const float w = uvw_w(bl, idx);

        const float phase_index = u * l + v * m_val + w * n;
        const float phase_offset =
            u_offset * l + v_offset * m_val + w_offset * n;

        for (size_t chan = channel_begin; chan < channel_end; ++chan) {
          const float phase = phase_offset - (phase_index * wavenumbers(chan));
          const std::complex<float> phasor =
              std::exp(std::complex<float>(0.0f, phase));

          for (size_t pol = 0; pol < nr_correlations_in; ++pol) {
            const size_t pol_out = pol % nr_correlations_out;
            pixels[pol_out] += visibilities(bl, idx, chan, pol) * phasor;
          }
        }
      }

      // Apply taper and store
      const float sph = taper(y, x);
      const size_t x_dst = (x + half_subgrid) % subgrid_size;
      const size_t y_dst = (y + half_subgrid) % subgrid_size;

      for (size_t pol = 0; pol < nr_correlations_out; ++pol) {
        subgrid(pol, y_dst, x_dst) = pixels[pol] * sph;
      }
    }
  }
}

xt::xarray<std::complex<float>> compute_phasor(int subgrid_size) {
  xt::xarray<std::complex<float>> phasor =
      xt::zeros<std::complex<float>>({subgrid_size, subgrid_size});

  for (int y = 0; y < subgrid_size; ++y) {
    for (int x = 0; x < subgrid_size; ++x) {
      float phase = M_PI * (x + y - subgrid_size) / subgrid_size;
      phasor(y, x) = std::exp(std::complex<float>(0.0f, phase));
    }
  }

  return phasor;
}

void add_subgrid_to_grid(int s, const std::vector<Metadata> &metadata,
                         const xt::xarray<std::complex<float>> &subgrids,
                         xt::xarray<std::complex<float>> &grid,
                         const xt::xarray<std::complex<float>> &phasor,
                         int nr_correlations, int subgrid_size, int grid_size) {
  const Metadata &m = metadata[s];
  int grid_x = m.coordinate.x;
  int grid_y = m.coordinate.y;

  if (grid_x >= 0 && grid_x < grid_size - subgrid_size && grid_y >= 0 &&
      grid_y < grid_size - subgrid_size) {

    for (int y = 0; y < subgrid_size; ++y) {
      for (int x = 0; x < subgrid_size; ++x) {
        int x_src = (x + subgrid_size / 2) % subgrid_size;
        int y_src = (y + subgrid_size / 2) % subgrid_size;

        for (int p = 0; p < nr_correlations; ++p) {
          grid(p, grid_y + y, grid_x + x) +=
              subgrids(s, p, y_src, x_src) * phasor(y, x);
        }
      }
    }
  }
}

void visibilities_to_subgrid(
    size_t s, const std::vector<Metadata> &metadata, float w_step,
    int grid_size, float image_size, const xt::xarray<float> &wavenumbers,
    const xt::xarray<std::complex<float>> &visibilities,
    const xt::xarray<UVW> &uvw, const xt::xarray<float> &taper,
    size_t nr_correlations_in, size_t subgrid_size,
    xt::xarray<std::complex<float>> &subgrid) {
  const Metadata &m = metadata[s];
  const int bl = m.baseline;
  const int offset = m.time_index;
  const int nr_timesteps = m.nr_timesteps;
  const int channel_begin = m.channel_begin;
  const int channel_end = m.channel_end;
  const int x_coordinate = m.coordinate.x;
  const int y_coordinate = m.coordinate.y;
  const float w_offset_in_lambda = w_step * (m.coordinate.z + 0.5f);

  const size_t nr_correlations_out = (nr_correlations_in == 4) ? 4 : 1;

  // Compute offsets
  const float u_offset =
      (x_coordinate + subgrid_size / 2.0f - grid_size / 2.0f) *
      (2.0f * M_PI / image_size);
  const float v_offset =
      (y_coordinate + subgrid_size / 2.0f - grid_size / 2.0f) *
      (2.0f * M_PI / image_size);
  const float w_offset = 2.0f * M_PI * w_offset_in_lambda;

  const float half_subgrid = subgrid_size / 2.0f;
  const float image_scale = image_size / subgrid_size;

  for (size_t y = 0; y < subgrid_size; ++y) {
    for (size_t x = 0; x < subgrid_size; ++x) {
      // Compute l, m, n
      const float l = (x + 0.5f - half_subgrid) * image_scale;
      const float m_val = (y + 0.5f - half_subgrid) * image_scale;
      const float tmp = l * l + m_val * m_val;
      const float n = tmp / (1.0f + std::sqrt(1.0f - tmp));

      // Compute pixels
      std::vector<std::complex<float>> pixels(nr_correlations_out,
                                              std::complex<float>(0.0f, 0.0f));

      for (size_t time = 0; time < nr_timesteps; ++time) {
        const int idx = offset + time;
        const float u = uvw(bl, idx).u;
        const float v = uvw(bl, idx).v;
        const float w = uvw(bl, idx).w;

        const float phase_index = u * l + v * m_val + w * n;
        const float phase_offset =
            u_offset * l + v_offset * m_val + w_offset * n;

        for (size_t chan = channel_begin; chan < channel_end; ++chan) {
          const float phase = phase_offset - (phase_index * wavenumbers(chan));
          std::complex<float> phasor =
              std::exp(std::complex<float>(0.0f, phase));

          for (size_t pol = 0; pol < nr_correlations_in; ++pol) {
            int pol_out = pol % nr_correlations_out;
            pixels[pol_out] += visibilities(bl, idx, chan, pol) * phasor;
          }
        }
      }

      // Apply taper and store
      const float sph = taper(y, x);
      const int x_dst = static_cast<int>((x + half_subgrid)) % subgrid_size;
      const int y_dst = static_cast<int>((y + half_subgrid)) % subgrid_size;

      for (size_t pol = 0; pol < nr_correlations_out; ++pol) {
        subgrid(pol, y_dst, x_dst) = pixels[pol] * sph;
      }
    }
  }
}